allennlp/training/learning_rate_schedulers.py

```diff
@@ -126,6 +126,100 @@ def get_lr(self):
+        self.num_epochs = num_epochs
+        self.num_steps_per_epoch = num_steps_per_epoch
+        self.cut_frac = cut_frac
+        self.ratio = ratio
+        self.lr = lr
+        self.gradual_unfreezing = gradual_unfreezing
+        self.freezing_current = self.gradual_unfreezing
+        self.first_epoch = True
+        super().__init__(optimizer, last_epoch=last_epoch)
+
+    def step(self, epoch=None):
+        if self.gradual_unfreezing:
+            assert not self.optimizer.param_groups[-1]["params"], \
+                "The default group should be empty."
+            # the method is called once when initialising before the
+            # first epoch (epoch 0) and then always at the end of each
+            # epoch; so the first time, with epoch id 0, we want to set
+            # up for epoch #1; the second time, still with epoch id 0,
+            # we want to set up for epoch #2, etc.
+            epoch_no = epoch+1 if self.first_epoch else epoch+2
+            if self.first_epoch:
+                self.first_epoch = False
+            if epoch_no >= len(self.optimizer.param_groups)-1:
+                print('Gradual unfreezing finished. Training all layers.')
+                self.freezing_current = False
+            else:
+                print(f'Gradual unfreezing. Training only the top {epoch_no} layers.')
+            for i, param_group in enumerate(reversed(self.optimizer.param_groups)):
+                for param in param_group["params"]:
+                    # i = 0 is the default group; we care about i > 0
+                    if i <= epoch_no:
+                        param.requires_grad = True
+                    else:
+                        param.requires_grad = False
+
+    def step_batch(self, epoch=None):
+        if epoch is None:
+            epoch = self.last_epoch + 1
+        self.last_epoch = epoch
+        for param_group, learning_rate in zip(self.optimizer.param_groups, self.get_lr()):
+            param_group['lr'] = learning_rate
+
+    def get_lr(self):
+        if self.freezing_current:
+            # if we still freeze, we restrict the schedule to the current epoch
+            num_steps = self.num_steps_per_epoch
+            step = max(self.last_epoch, 1) % self.num_steps_per_epoch
+        else:
+            # otherwise we use the schedule for the rest of training
+            steps_while_frozen = (len(self.optimizer.param_groups)-1) * self.num_steps_per_epoch
+            num_steps = self.num_epochs * self.num_steps_per_epoch - steps_while_frozen
+            step = max(self.last_epoch, 1) - steps_while_frozen
+        cut = int(num_steps * self.cut_frac)
+        if cut == step:
+            print()
+        p = step / cut if step < cut else 1 - (step - cut) / (num_steps - cut)
+        lrs = [lr * (1 + p * (self.ratio - 1)) / self.ratio for lr in self.base_lrs]
+        return lrs
+
 # We just use the Pytorch LRSchedulers, so here we force them into
```

Comment:
This is maybe a bit annoying for you to have to do @sebastianruder, so we can just open an issue instead of fixing this in this PR, but it seems like we've got enough learning rate schedulers that we should separate this class out into a module, and give `NoamLR` and your scheduler their own files.  Then the built-in pytorch schedulers can still live here in this dictionary, but the classes we implement can just use `@LearningRateScheduler.register()` as normal.

Reply:
Cool. Let me know if you'd like me to do this here or if you prefer to create a separate issue.

Reply:
Whether it's you that does it or someone else, it's probably better done in a separate PR.